### DISCUSSION TITLE
When loading from a TAP file, should the pointer wrap around to the beginning or issue an error?

### DISCUSSION THREAD

**Discussion by factus10** [timex-sinclair-projects/TS-Pico#8] [Ideas]
Kevin: "LOADing a program, I think that there should be an error generated once a TS-Pico reaches the end of a tap(e)."

**Comment by BetaTesterK** [comment]
The error generated could be a 1 line BASIC program similar to the one loaded already for another error.  For Example:

1 REM Reached the end of the .TAP file without finding anything to load.

- Kevin

**Comment by rcalcagno** [comment]
Actually, to not complicate matters further, the same fake program can be used for both situations, and will only require a single line of code. The program could looke like ths:

10 REM No .TAP file mounted
20 REM Or end of .TAP reached
30 REM Please (re)mount a .TAP file
40 REM and try again.

What people think about this?

**Comment by BetaTesterK** [comment]
I think it's a great idea.  It needs to be reformatted a little to fit
the 32 column mode:

10 REM No .TAP file mounted or 
20 REM end of .TAP reached.
30 REM Please (re)mount a .TAP
40 REM file and try again.

Alternatively, you could format this as a 1 line REM that would look
good in 32 and 64 columns:

---64 COL mode:

  10 REM No .TAP file mounted   -or- end of .TAP file reached.  
Please (re)mount a .TAP file and try again.

---32 Col mode:

  10 REM No .TAP file mounted   
-or- end of .TAP file reached.  
Please (re)mount a .TAP file and
 try again.

I'd prefer the fake program generated be smart enough to detect which
problem had occurred, but this would work fine if that's not possible.
Also, this could be the quick fix to get this version finished and into
production, leaving a "smarter" program generator for the next version.

One final suggestion:

What if this fake program was a PRINT instead of a REM statement?  You
could have the program auto run (e. g. as if SAVE "Fake Prgm" LINE 10)
and actually print the error on the screen.  Maybe this should be a
config option.

**Comment by rcalcagno** [comment]
Well, actually it's not a matter of complexity; it's almost trivial to add a second "fake" .TAP that carries the .TAP EOF message instead of the generic one; it's more a matter (for me) of "unifying" the problem at the user's eye. IMHO, a single message for both situations is a typical case of less-is-more; but again, this is opinion, not a fact. And actually shows how rich and important is this chance to try and discuss several methods, and not just implement what one single person believes fits the bill better.

Another thing to take into account is that, as it stands right now, a LOAD "foo" will always fail, once we went thru all the headers on the .TAP file no matter what strategy we implement (either silently rewind or load the fake .TAP), since the TS will keep asking for more headers until it reaches the (incorrect) one we asked for. There are two ways we can escape from this, if we do not auto-rewind:

a) The Pico could NOT send anything else after the last header on a .TAP file, not even the fake one; and resume waiting for new commands. In this case, the user himself will need to hit CTRL+BREAK to exit the LOAD loop (as he/she will in the case of a real tape); or
b) The Pico could send a scrambled CRC and induce an error (I believe this was KEVIN'S suggestion). I personally believe this is less preferably, as one would never know whether a "real" CRC error has occurred, or we've reached the .TAP's end. No additional error info will be available as in the case of the "J Error" situation, only the log will keep what really happened.
